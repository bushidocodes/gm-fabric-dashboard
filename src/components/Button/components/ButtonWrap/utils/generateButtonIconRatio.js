/**
 * Helper function that maps button sizes to numbers representing icon ratios
 *
 * @param {string} size
 * @returns number
 */
function mapSizeToButtonIconRatio(size) {
  switch (size) {
    case "xs":
      return 0.5;
    case "sm":
      return 0.75;
    case "lg":
      return 1.25;
    case "xl":
      return 1.5;
    case "normal":
    default:
      return 1;
  }
}

/**
 * Utility function that generates a CSS mixin containing button icon rulesets
 *
 * @param {string} size
 * @returns string
 */
function generateButtonIconRatio(size) {
  const viewboxSize = 24; // default viewbox size 24px
  const ratio = mapSizeToButtonIconRatio(size); // default ratio 1
  const viewboxDimensions = viewboxSize * ratio;
  return `> svg { width: ${/* sc-value */ viewboxDimensions}px; 
  height: ${/* sc-value */ viewboxDimensions}px; }`;
}

export default generateButtonIconRatio;
