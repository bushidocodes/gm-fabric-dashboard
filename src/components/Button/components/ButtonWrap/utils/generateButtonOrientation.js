import { spacingScale } from "style/styleFunctions";

/**
 * Utility function that generates a CSS mixin containing button orientation rulesets
 *
 * @param {string} [alignment="horizontal"]
 * @param {number} [spacingBase=0.5]
 * @returns string
 */
function generateButtonOrientation(
  alignment = "horizontal",
  spacingBase = 0.5
) {
  if (alignment === "horizontal") {
    return `
      flex-direction: row;
      > :not(:first-child) {
        margin-left: ${/* sc-value */ spacingScale(spacingBase)};
        flex-basis: auto;
      }
    `;
  } else if (alignment === "vertical") {
    return `
      flex-direction: column;
      padding-top: ${/* sc-value */ spacingScale(spacingBase)};
      > :not(:first-child) {
        margin-top: ${/* sc-value */ spacingBase / 2}px;
      }
    `;
  }
}

export default generateButtonOrientation;
