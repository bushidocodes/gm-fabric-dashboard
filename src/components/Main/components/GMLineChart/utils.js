// importing a single function instead of the whole _
import isEmpty from "lodash/isEmpty";

/**
 * Converts timeseries with time stamps to a string of values without timestamps
 *
 * @param {array} array a timeseries array with time stamps
 * @returns {array}     an array of values
 */
export function numericalTimeSeriesFunc(array) {
  if (array instanceof Array === true && !isEmpty(array)) {
    return array[0].map(cell => cell[1]);
  } else {
    return 0;
  }
}

/**
 * Produces an average of array of values
 *
 * @param {array} array an array of values
 * @returns {number}
 */
export function average(array) {
  if (array instanceof Array === true && !isEmpty(array)) {
    return array.reduce((a, b) => a + b) / array.length;
  } else {
    return 0;
  }
}

/**
 * Produces a median of array of values
 * (if no middle value, average of 2 most central values is returned)
 *
 * @param {array} array an array of values
 * @returns {number}
 */
export function median(array) {
  if (array instanceof Array === true && !isEmpty(array)) {
    const len = array.length;
    let median = 0;
    array.sort();

    if (len % 2 === 0) {
      // if even, average of two middle numbers
      median = (array[len / 2 - 1] + array[len / 2]) / 2;
    } else {
      // if odd, middle number only
      median = array[(len - 1) / 2];
    }
    return median;
  } else {
    return 0;
  }
}

/**
 * Produces a mode(s) of array of values
 * if more than one mode, returns all applicable modes as well as their frequency
 *
 * @param {array} array an array of values
 * @returns {array}     an array that contains mode(s) and frequency value
 */
export function modes(array) {
  if (array instanceof Array === true && !isEmpty(array)) {
    let modes = [];
    let count = [];
    let i;
    let number;
    let maxIndex = 0;

    for (let i = 0; i < array.length; i++) {
      number = array[i];
      count[number] = (count[number] || 0) + 1;
      if (count[number] > maxIndex) {
        maxIndex = count[number];
      }
    }

    for (i in count)
      if (count.hasOwnProperty(i)) {
        if (count[i] === maxIndex) {
          modes.push(Number(i));
        }
      }

    modes.push({ frequency: maxIndex });

    if (maxIndex > 1) {
      return modes;
    } else {
      modes = "no value occurs more than once";
      return modes;
    }
  } else {
    return 0;
  }
}

/**
 * Produces a string that contains descriptive statistics of the graph
 *
 * @param {array} timeSeries an array of values
 * @param {string} title chart title
 * @returns {string}     returns a string with summary chart statistics
 */
export function screenReaderGraphDescription(timeSeries, title, intl) {
  const numericalTimeSeries = numericalTimeSeriesFunc(timeSeries);

  // We don't have negative values in our charts, thus average of 0 means that all values are 0, aka no meaningful data
  if (average(numericalTimeSeries) > 0) {
    let screenReaderGraphData = {
      median: median(numericalTimeSeries).toFixed(2),
      average: average(numericalTimeSeries).toFixed(2),
      mode: modes(numericalTimeSeries),
      max: numericalTimeSeries.sort()[numericalTimeSeries.length - 1],
      min: numericalTimeSeries.sort()[0],
      dataPoints: numericalTimeSeries.length,
      timeSeries: numericalTimeSeries,
      title
    };

    if (
      typeof screenReaderGraphData.mode !== "string" &&
      screenReaderGraphData.mode.length > 1
    ) {
      screenReaderGraphData = Object.assign({}, screenReaderGraphData, {
        mode: `Mode has frequency of ${
          screenReaderGraphData.mode.pop().frequency
        } and values of ${screenReaderGraphData.mode}`
      });
    }

    return intl.formatMessage(
      {
        id: "GMLineChart.screenReaderGraphDescription.withData",
        defaultMessage:
          "A tabular representation of the {title} chart data: median {median} average {average} mode {mode} maximum {max} minimum {min} number of observations {dataPoints} complete data time series follows {timeSeries}",
        description: "Screen reader description for GMLineChart"
      },
      screenReaderGraphData
    );
  } else {
    return intl.formatMessage({
      id: "GMLineChart.screenReaderGraphDescription.noData",
      defaultMessage: `The average for currently displayed data is equal to 0.`,
      description: "Screen reader description for GMLineChart"
    });
  }
}
