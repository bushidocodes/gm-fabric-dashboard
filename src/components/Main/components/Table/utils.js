import React from "react";
import TableColHeader from "components/Main/components/TableColHeader";
import TableColLatencyHeader from "components/Main/components/TableColLatencyHeader";

/** function getTableHeaders
 * takes Table type and returns appropriate headers with styling
 * for Function/Routes Table or GMServiceTable (Instances Table)
 * @param {object} items
 * @returns {React.element}
 */
export const getTableHeaders = type => {
  let headerNames = [];
  if (type === "Instance") headerNames = [type, "Uptime"];
  if (type === "Route" || type === "Function")
    headerNames = [type, "Requests/sec", "Requests", "Error %", "Latency"];
  let styleProps = {};

  // apply greater flex-basis for first column (20%)
  // requests/sec is a sparkline header and set to align left
  // set text-align to right for additional text columns
  return headerNames.map((header, headerIdx) => {
    styleProps = {};
    if (headerIdx === 0) styleProps = { flex: "1 1 20%" };
    if (headerIdx !== 0 && header !== "Requests/sec")
      styleProps = { textAlign: "right" };

    if (header === "Latency") {
      return <TableColLatencyHeader key={header} />;
    } else {
      return (
        <TableColHeader style={styleProps} key={header}>
          {header}
        </TableColHeader>
      );
    }
  });
};

// simple helper that evaluates and returns route or func props
export const getItem = (route, func) => {
  if (route !== undefined) return route;
  if (func !== undefined) return func;
};
