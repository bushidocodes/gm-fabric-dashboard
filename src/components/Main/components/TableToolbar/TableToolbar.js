// NPM Modules
import React from "react";
import { PropTypes } from "prop-types";
import { injectIntl, intlShape } from "react-intl";
// Internal Components
import SearchInput from "components/Main/components/GMSearchInput";
import NavTab from "components/NavTab";
import Form from "./components/Form";
import GMSelect from "components/Main/components/GMSelect";
import { GMSelectValueRenderer } from "components/Main/components/GMSelect";
import Toolbar from "./components/Toolbar";
import ToolbarLeft from "./components/ToolbarLeft";
import ToolbarRight from "./components/ToolbarRight";
import ToolbarCenter from "./components/ToolbarCenter";

/** A reusable toolbar that renders any combination of search box, display type toggle buttons,
 *  and sorting or grouping dropdowns. To render a subcomponent, pass it the correct props.
 *  You may also use toolbar children props to render custom subcomponents.
 * @function TableToolbar
 */

TableToolbar.propTypes = {
  displayTypeProps: PropTypes.shape({
    displayType: PropTypes.oneOf(["Cards", "List"]).isRequired, // if this is a Fabric Grid Toolbar, display type buttons will be rendered
    setDisplayType: PropTypes.func.isRequired // onClick handler for ToolbarCenter buttons. Takes in either “Card” or “Table”.
  }),
  groupByProps: PropTypes.shape({
    groupByAttribute: PropTypes.string.isRequired, // value for controlled GMSelect Group By dropdown
    groupByOptions: PropTypes.array.isRequired, // array of option objects for the GMSelect Group By dropdown
    setGroupByAttribute: PropTypes.func.isRequired // onChange handler for GMSelect Group By dropdown
  }),
  intl: intlShape.isRequired,
  searchInputProps: PropTypes.shape({
    filterString: PropTypes.string.isRequired, // value for controlled SearchInput input
    setFilterString: PropTypes.func.isRequired, // onChange handler for search input
    searchPlaceholder: PropTypes.string.isRequired // placeholder for SearchInput
  }),
  sortByProps: PropTypes.shape({
    setSortByAttribute: PropTypes.func, // onChange handler for GMSelect Sort By dropdown
    sortByAttribute: PropTypes.string, // value for controlled GMSelect Sort By dropdown
    sortByOptions: PropTypes.array // array of option objects for the GMSelect Sort By dropdown
  }),
  // the following are optional props to be rendered as children of their respective columns
  toolbarCenterChildren: PropTypes.oneOfType([
    PropTypes.element,
    PropTypes.node
  ]),
  toolbarLeftChildren: PropTypes.oneOfType([PropTypes.element, PropTypes.node]),
  toolbarRightChildren: PropTypes.oneOfType([PropTypes.element, PropTypes.node])
};

function TableToolbar({
  displayTypeProps,
  searchInputProps,
  groupByProps,
  sortByProps,
  toolbarLeftChildren,
  toolbarCenterChildren,
  toolbarRightChildren,
  intl
}) {
  return (
    <Toolbar>
      <ToolbarLeft>
        {searchInputProps && (
          <Form>
            <SearchInput
              onChange={evt =>
                searchInputProps.setFilterString(evt.target.value)
              }
              placeholder={searchInputProps.searchPlaceholder}
              aria-label={searchInputProps.searchPlaceholder}
              value={searchInputProps.filterString}
            />
          </Form>
        )}
        {toolbarLeftChildren}
      </ToolbarLeft>
      <ToolbarCenter>
        {displayTypeProps && [
          <NavTab
            active={displayTypeProps.displayType === "Cards"}
            clickAction={() => displayTypeProps.setDisplayType("Cards")}
            glyph="Card"
            label={intl.formatMessage({
              id: "tableToolbar.cards",
              defaultMessage: "Cards",
              description: "Display type button"
            })}
            key="Cards"
          />,
          <NavTab
            active={displayTypeProps.displayType === "List"}
            clickAction={() => displayTypeProps.setDisplayType("List")}
            glyph="List"
            label={intl.formatMessage({
              id: "tableToolbar.list",
              defaultMessage: "List",
              description: "Display type button"
            })}
            key="List"
          />
        ]}
        {toolbarCenterChildren}
      </ToolbarCenter>
      <ToolbarRight>
        {groupByProps && (
          <GMSelect
            matchPos="start"
            name="form-field-group-by"
            options={groupByProps.groupByOptions}
            value={groupByProps.groupByAttribute}
            onChange={val => {
              val && groupByProps.setGroupByAttribute(val.value);
            }}
            clearable={false}
            searchable={true}
            valueRenderer={val => (
              <GMSelectValueRenderer
                title={intl.formatMessage({
                  id: "tableToolbar.group",
                  defaultMessage: "Group",
                  description: "Group by dropdown title"
                })}
                val={val}
              />
            )}
          />
        )}
        {sortByProps && (
          <GMSelect
            matchPos="start"
            name="form-field-sort-by"
            options={sortByProps.sortByOptions}
            value={sortByProps.sortByAttribute}
            onChange={val => {
              val && sortByProps.setSortByAttribute(val.value);
            }}
            clearable={false}
            searchable={true}
            valueRenderer={val => (
              <GMSelectValueRenderer
                title={intl.formatMessage({
                  id: "tableToolbar.sort",
                  defaultMessage: "Sort",
                  description: "Sort by dropdown title"
                })}
                val={val}
              />
            )}
          />
        )}
        {toolbarRightChildren}
      </ToolbarRight>
    </Toolbar>
  );
}

export default injectIntl(TableToolbar);
