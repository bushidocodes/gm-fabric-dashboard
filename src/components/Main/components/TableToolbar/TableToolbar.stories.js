import React from "react";
import { storiesOf } from "@storybook/react";

import TableToolbar from "./index.js";

const sortByOptions = [
  {
    value: "Name",
    label: "Name"
  },
  {
    value: "Status",
    label: "Status"
  }
];

const groupByOptions = [
  {
    value: "Owner",
    label: "Owner"
  },
  {
    value: "Status",
    label: "Status"
  }
];

const mockFabricViewProps = {
  displayTypeProps: {
    displayType: "Cards",
    setDisplayType: () => alert("fired setDisplayType")
  },
  searchInputProps: {
    filterString: "",
    searchPlaceholder: "Search Services",
    setFilterString: () => alert("fired setFilterString")
  },
  groupByProps: {
    groupByOptions,
    groupByAttribute: "Status",
    setGroupByAttribute: () => alert("fired setGroupByAttribute")
  },
  sortByProps: {
    sortByOptions,
    sortByAttribute: "Name",
    setSortByAttribute: () => alert("fired setSortByAttribute")
  }
};

storiesOf("Table Toolbar", module).add("default", () => {
  return (
    <div style={{ width: "100%" }}>
      <TableToolbar {...mockFabricViewProps} />
    </div>
  );
});
