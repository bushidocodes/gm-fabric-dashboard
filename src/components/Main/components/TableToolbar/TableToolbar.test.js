import React from "react";
import { mountWithIntl, renderWithIntl } from "utils/i18nTesting";

import TableToolbar from "./index.js";
import NavTab from "components/NavTab";

const sortByOptions = [
  {
    value: "Name",
    label: "Name"
  },
  {
    value: "Status",
    label: "Status"
  }
];

const groupByOptions = [
  {
    value: "Owner",
    label: "Owner"
  },
  {
    value: "Status",
    label: "Status"
  }
];

const mockFabricViewProps = {
  displayTypeProps: {
    displayType: "Cards",
    setDisplayType: jest.fn()
  },
  searchInputProps: {
    filterString: "",
    setFilterString: jest.fn(),
    searchPlaceholder: "Search Services"
  },
  groupByProps: {
    groupByOptions,
    groupByAttribute: "Status",
    setGroupByAttribute: jest.fn()
  },
  sortByProps: {
    sortByOptions,
    sortByAttribute: "Name",
    setSortByAttribute: jest.fn()
  }
};

describe("Table Toolbar", () => {
  let wrapper;

  beforeEach(() => {
    wrapper = mountWithIntl(<TableToolbar {...mockFabricViewProps} />);
  });

  test("matches snapshot", () => {
    const tree = renderWithIntl(<TableToolbar {...mockFabricViewProps} />);
    expect(tree).toMatchSnapshot();
  });

  test("renders search box", () => {
    expect(wrapper.find({ type: "search" })).toHaveLength(1);
  });

  test("calls setFilterString when search input field changes", () => {
    wrapper.find({ type: "search" }).simulate("change");
    expect(wrapper.props().searchInputProps.setFilterString).toHaveBeenCalled();
  });

  test("does not render search box if searchInputProps is not provided", () => {
    wrapper.setProps({
      searchInputProps: null
    });
    expect(wrapper.find({ type: "search" })).toHaveLength(0);
  });

  test("renders display type tabs", () => {
    // Find tabs with title attributes "Cards" and "List"
    expect(wrapper.find("a").find({ title: "Cards" })).toHaveLength(1);
    expect(wrapper.find("a").find({ title: "List" })).toHaveLength(1);
  });

  test("calls setDisplayType when a display type tab is clicked", () => {
    // Simulate a click and check if setDisplayType was called
    wrapper
      .find("a")
      .find({ title: "Cards" })
      .simulate("click");
    expect(wrapper.props().displayTypeProps.setDisplayType).toHaveBeenCalled();
  });

  test("does not render display type tabs if displayTypeProps is not provided", () => {
    wrapper.setProps({
      displayTypeProps: null
    });
    // Find tabs with title attributes "Cards" and "List"
    expect(wrapper.find("a").find({ title: "Cards" })).toHaveLength(0);
    expect(wrapper.find("a").find({ title: "List" })).toHaveLength(0);
  });

  test("adds an active class to tab that matches displayType", () => {
    let cardTab = wrapper.find(NavTab).at(0);
    let listTab = wrapper.find(NavTab).at(1);
    expect(cardTab.props().active).toBe(true);
    expect(listTab.props().active).toBe(false);
    wrapper.setProps({
      displayTypeProps: { displayType: "List", setDisplayType: jest.fn() }
    });
    cardTab = wrapper.find(NavTab).at(0);
    listTab = wrapper.find(NavTab).at(1);
    expect(listTab.props().active).toBe(true);
    expect(cardTab.props().active).toBe(false);
  });

  // TODO: Figure out how to simulate a change event with react-select so we can test onChange handlers

  test("renders a group by dropdown with a value equal to groupByAttribute", () => {
    const dropdown = wrapper
      .find("Select")
      .find("input")
      .find({ name: "form-field-group-by" });
    expect(dropdown).toHaveLength(1);
    expect(dropdown.props().value).toBe("Status");
  });

  test("does not render a group by dropdown if groupByProps is not provided", () => {
    wrapper.setProps({
      groupByProps: null
    });
    const dropdown = wrapper
      .find("Select")
      .find("input")
      .find({ name: "form-field-group-by" });
    expect(dropdown).toHaveLength(0);
  });

  test("renders a sort by dropdown", () => {
    const dropdown = wrapper
      .find("Select")
      .find("input")
      .find({ name: "form-field-sort-by" });
    expect(dropdown).toHaveLength(1);
    expect(dropdown.props().value).toBe("Name");
  });

  test("does not render a group by dropdown if sortByProps is not provided", () => {
    wrapper.setProps({
      sortByProps: null
    });
    const dropdown = wrapper
      .find("Select")
      .find("input")
      .find({ name: "form-field-sort-by" });
    expect(dropdown).toHaveLength(0);
  });

  test("optional children props render additional nodes into their corresponding columns", () => {
    wrapper.setProps({
      toolbarLeftChildren: <h1>Left Child</h1>,
      toolbarCenterChildren: <h1>Center Child</h1>,
      toolbarRightChildren: <h1>Right Child</h1>
    });
    const headers = wrapper.find("h1");
    expect(headers).toHaveLength(3);
    expect(headers.at(0).text()).toBe("Left Child");
    expect(headers.at(1).text()).toBe("Center Child");
    expect(headers.at(2).text()).toBe("Right Child");
  });
});
