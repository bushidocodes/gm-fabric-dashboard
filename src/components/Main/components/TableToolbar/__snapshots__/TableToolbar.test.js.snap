// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table Toolbar matches snapshot 1`] = `
.c2 {
  background-image: url(magnifier.svg);
  background-repeat: no-repeat;
  background-size: contain;
  position: relative;
  text-indent: 20px;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  border: 1px solid rgba(0,0,0,0.25);
  border-radius: 3px;
  text-align: left;
  background-color: model:rgb;
  color: 255,255,255;
  valpha: 1;
  font-size: 14px;
  font-weight: 400;
  line-height: 1.1;
  width: 22em;
  -webkit-flex: 1 1 22em;
  -ms-flex: 1 1 22em;
  flex: 1 1 22em;
  padding: 6px 8px;
}

.c2::-webkit-search-cancel-button {
  cursor: pointer;
}

.c2::-webkit-input-placeholder {
  color: inherit;
  opacity: 1;
}

.c2::-moz-placeholder {
  color: inherit;
  opacity: 1;
}

.c2:-ms-input-placeholder {
  color: inherit;
  opacity: 1;
}

.c2::placeholder {
  color: inherit;
  opacity: 1;
}

.c7 {
  -webkit-transform: scale(1);
  -ms-transform: scale(1);
  transform: scale(1);
}

.c6 {
  z-index: 1;
  height: 24px;
  width: 24px;
  color: currentColor;
  word-spacing: 0;
  vertical-align: middle;
}

.c6:not(:root) {
  overflow: visible;
}

.c5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  padding: 8px 16px;
  -webkit-flex: 0 1 auto;
  -ms-flex: 0 1 auto;
  flex: 0 1 auto;
  cursor: pointer;
  line-height: 0;
  -webkit-transition: all 0.15s ease;
  transition: all 0.15s ease;
}

.c5:hover {
  box-shadow: inset 0 -1px rgba(128,128,128,0.5);
}

.c5:hover:active {
  box-shadow: inset 0 -1.5px rgb(128,128,128);
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c5,
.c5:hover {
  box-shadow: inset 0 -2px rgb(10,171,42);
}

.c5:active,
.c5:hover:active {
  box-shadow: inset 0 -2px hsl(131.89999999999998,89%,28.4%);
}

.c5 > svg:first-child,
.c5:hover > svg:first-child {
  margin-left: -4px;
}

.c5 > svg:last-child,
.c5:hover > svg:last-child {
  margin-right: -4px;
}

.c8 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  padding: 8px 16px;
  -webkit-flex: 0 1 auto;
  -ms-flex: 0 1 auto;
  flex: 0 1 auto;
  cursor: pointer;
  line-height: 0;
  -webkit-transition: all 0.15s ease;
  transition: all 0.15s ease;
}

.c8:hover {
  box-shadow: inset 0 -1px rgba(128,128,128,0.5);
}

.c8:hover:active {
  box-shadow: inset 0 -1.5px rgb(128,128,128);
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c8 > svg:first-child {
  margin-left: -4px;
}

.c8 > svg:last-child {
  margin-right: -4px;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  min-height: 40px;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
}

.c10 {
  -webkit-flex-grow: 1;
  -ms-flex-grow: 1;
  flex-grow: 1;
  font-size: 14px;
  font-weight: 500;
  height: 28px;
  max-width: 125px;
  position: relative;
  width: 100%;
  z-index: 1020;
}

.c10 .Select-control {
  height: 28px !important;
}

.c10 .Select-control .Select-placeholder {
  line-height: 28px;
}

.c10 .Select-control .Select-value {
  line-height: 28px !important;
}

.c10 .Select-control .Select-value .Select-value-label {
  line-height: 28px;
}

.c10 .Select-control .Select-input {
  height: 28px;
}

.c10 + .c10 {
  margin-left: 8px;
}

.c11 {
  margin-left: 4px;
  opacity: 0.6;
  font-size: 14px;
}

.c0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: hsl(0,0%,96%);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  height: 36px;
  margin-left: auto;
  margin-right: auto;
  width: 100%;
  max-width: 1300px;
  padding: 2px 16px;
}

.c1 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1 1 33%;
  -ms-flex: 1 1 33%;
  flex: 1 1 33%;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
}

.c9 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1 1 33%;
  -ms-flex: 1 1 33%;
  flex: 1 1 33%;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1 1 33%;
  -ms-flex: 1 1 33%;
  flex: 1 1 33%;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

@media all and (min-width:1300px) {
  .c0:after {
    background-color: inherit;
    content: "";
    height: inherit;
    left: 0;
    pointer-events: none;
    position: absolute;
    right: 0;
    top: 0;
    z-index: 1;
  }

  .c0 > * {
    position: relative;
    z-index: 2;
  }
}

<div
  class="c0"
>
  <div
    class="c1"
  >
    <form>
      <input
        aria-label="Search Services"
        class="c2"
        placeholder="Search Services"
        type="search"
        value=""
      />
    </form>
  </div>
  <div
    class="c3"
  >
    <nav
      class="c4"
    >
      <a
        class="c5"
        title="Cards"
      >
        <svg
          aria-labelledby="ariaLabelledby"
          class="c6"
          focusable="false"
          version="1.1"
          viewBox="0 0 24 24"
          xlink="http://www.w3.org/1999/xlink"
          xmlns="http://www.w3.org/2000/svg"
        >
          <g
            class="c7"
            fill="currentColor"
            title=""
          >
            <title />
            <g
              class="glyph"
              fill="currentColor"
              transform="translate(0 0) scale(1)"
            >
              <path
                d="M7,4 L17,4 C18.1045695,4 19,4.8954305 19,6 L19,17 L19,17 C19,18.1045695 18.1045695,19 17,19 L7,19 L7,19 C5.8954305,19 5,18.1045695 5,17 L5,6 L5,6 C5,4.8954305 5.8954305,4 7,4 L7,4 Z M10.5,16 L16.5,16 C16.7761424,16 17,16.2238576 17,16.5 C17,16.7761424 16.7761424,17 16.5,17 L10.5,17 C10.2238576,17 10,16.7761424 10,16.5 C10,16.2238576 10.2238576,16 10.5,16 Z M7.5,10 L14.5,10 C14.7761424,10 15,10.2238576 15,10.5 C15,10.7761424 14.7761424,11 14.5,11 L7.5,11 C7.22385763,11 7,10.7761424 7,10.5 C7,10.2238576 7.22385763,10 7.5,10 Z M7.5,12 L10.5,12 C10.7761424,12 11,12.2238576 11,12.5 C11,12.7761424 10.7761424,13 10.5,13 L7.5,13 C7.22385763,13 7,12.7761424 7,12.5 C7,12.2238576 7.22385763,12 7.5,12 Z M8,6 L16,6 C16.5522847,6 17,6.44771525 17,7 L17,8 C17,8.55228475 16.5522847,9 16,9 L8,9 C7.44771525,9 7,8.55228475 7,8 L7,7 C7,6.44771525 7.44771525,6 8,6 Z M7,5 L7,5 C6.44771525,5 6,5.44771525 6,6 L6,17 C6,17.5522847 6.44771525,18 7,18 L17,18 C17.5522847,18 18,17.5522847 18,17 L18,6 C18,5.44771525 17.5522847,5 17,5 L7,5 Z"
                id="path-1"
              />
            </g>
          </g>
        </svg>
        <span>
          Cards
        </span>
      </a>
      <a
        class="c8"
        title="List"
      >
        <svg
          aria-labelledby="ariaLabelledby"
          class="c6"
          focusable="false"
          version="1.1"
          viewBox="0 0 24 24"
          xlink="http://www.w3.org/1999/xlink"
          xmlns="http://www.w3.org/2000/svg"
        >
          <g
            class="c7"
            fill="currentColor"
            title=""
          >
            <title />
            <g
              class="glyph"
              fill="currentColor"
              transform="translate(0 0) scale(1)"
            >
              <path
                d="M5.5,8 C4.67157288,8 4,7.32842712 4,6.5 C4,5.67157288 4.67157288,5 5.5,5 C6.32842712,5 7,5.67157288 7,6.5 C7,7.32842712 6.32842712,8 5.5,8 Z M5.5,13 C4.67157288,13 4,12.3284271 4,11.5 C4,10.6715729 4.67157288,10 5.5,10 C6.32842712,10 7,10.6715729 7,11.5 C7,12.3284271 6.32842712,13 5.5,13 Z M5.5,18 C4.67157288,18 4,17.3284271 4,16.5 C4,15.6715729 4.67157288,15 5.5,15 C6.32842712,15 7,15.6715729 7,16.5 C7,17.3284271 6.32842712,18 5.5,18 Z M9,5.5 L19,5.5 C19.5522847,5.5 20,5.94771525 20,6.5 L20,6.5 C20,7.05228475 19.5522847,7.5 19,7.5 L9,7.5 C8.44771525,7.5 8,7.05228475 8,6.5 L8,6.5 L8,6.5 C8,5.94771525 8.44771525,5.5 9,5.5 Z M9,10.5 L19,10.5 C19.5522847,10.5 20,10.9477153 20,11.5 C20,12.0522847 19.5522847,12.5 19,12.5 L9,12.5 C8.44771525,12.5 8,12.0522847 8,11.5 C8,10.9477153 8.44771525,10.5 9,10.5 Z M9,15.5 L19,15.5 C19.5522847,15.5 20,15.9477153 20,16.5 C20,17.0522847 19.5522847,17.5 19,17.5 L9,17.5 C8.44771525,17.5 8,17.0522847 8,16.5 C8,15.9477153 8.44771525,15.5 9,15.5 Z"
                id="Combined-Shape"
              />
            </g>
          </g>
        </svg>
        <span>
          List
        </span>
      </a>
    </nav>
  </div>
  <div
    class="c9"
  >
    <div
      class="Select c10 has-value is-searchable Select--single"
    >
      <input
        name="form-field-group-by"
        type="hidden"
        value="Status"
      />
      <div
        class="Select-control"
      >
        <span
          class="Select-multi-value-wrapper"
          id="react-select-2--value"
        >
          <div
            class="Select-value"
          >
            <span
              aria-selected="true"
              class="Select-value-label"
              id="react-select-2--value-item"
              role="option"
            >
              <span>
                <span>
                  Group 
                </span>
                <span
                  class="c11"
                >
                  Status
                </span>
              </span>
            </span>
          </div>
          <div
            class="Select-input"
            style="display:inline-block"
          >
            <input
              aria-activedescendant="react-select-2--value"
              aria-expanded="false"
              aria-haspopup="false"
              aria-owns=""
              role="combobox"
              style="box-sizing:content-box;width:5px"
              value=""
            />
            <div
              style="position:absolute;top:0;left:0;visibility:hidden;height:0;overflow:scroll;white-space:pre"
            />
          </div>
        </span>
        <span
          class="Select-arrow-zone"
        >
          <span
            class="Select-arrow"
          />
        </span>
      </div>
    </div>
    <div
      class="Select c10 has-value is-searchable Select--single"
    >
      <input
        name="form-field-sort-by"
        type="hidden"
        value="Name"
      />
      <div
        class="Select-control"
      >
        <span
          class="Select-multi-value-wrapper"
          id="react-select-3--value"
        >
          <div
            class="Select-value"
          >
            <span
              aria-selected="true"
              class="Select-value-label"
              id="react-select-3--value-item"
              role="option"
            >
              <span>
                <span>
                  Sort 
                </span>
                <span
                  class="c11"
                >
                  Name
                </span>
              </span>
            </span>
          </div>
          <div
            class="Select-input"
            style="display:inline-block"
          >
            <input
              aria-activedescendant="react-select-3--value"
              aria-expanded="false"
              aria-haspopup="false"
              aria-owns=""
              role="combobox"
              style="box-sizing:content-box;width:5px"
              value=""
            />
            <div
              style="position:absolute;top:0;left:0;visibility:hidden;height:0;overflow:scroll;white-space:pre"
            />
          </div>
        </span>
        <span
          class="Select-arrow-zone"
        >
          <span
            class="Select-arrow"
          />
        </span>
      </div>
    </div>
  </div>
</div>
`;
